[core]: add tool calls message

baskaryan opened the pull request:
```python
class ToolCall(TypedDict):
    name: str
    args: Dict[str, Any]
    id: Optional[str]

class InvalidToolCall(TypedDict):
    name: Optional[str]
    args: Optional[str]
    id: Optional[str]
    error: Optional[str]

class ToolCallChunk(TypedDict):
    name: Optional[str]
    args: Optional[str]
    id: Optional[str]
    index: Optional[int]


class AIMessage(BaseMessage):
    ...
    tool_calls: List[ToolCall] = []
    invalid_tool_calls: List[InvalidToolCall] = []
    ...


class AIMessageChunk(AIMessage, BaseMessageChunk):
    ...
    tool_call_chunks: Optional[List[ToolCallChunk]] = None
    ...
```
Important considerations:
- Parsing logic occurs within different providers;
- ~Changing output type is a breaking change for anyone doing explicit type checking;~
- ~Langsmith rendering will need to be updated: https://github.com/langchain-ai/langchainplus/pull/3561~
- ~Langserve will need to be updated~
- Adding chunks:
  - ~AIMessage + ToolCallsMessage = ToolCallsMessage if either has non-null .tool_calls.~
  - Tool call chunks are appended, merging when having equal values of `index`.
  - additional_kwargs accumulate the normal way.
- During streaming:
  - ~Messages can change types (e.g., from AIMessageChunk to AIToolCallsMessageChunk)~
  - Output parsers parse additional_kwargs (during .invoke they read off tool calls).

Packages outside of `partners/`:
- https://github.com/langchain-ai/langchain-cohere/pull/7
- https://github.com/langchain-ai/langchain-google/pull/123/files

baskaryan committed:
rfc: tool calls msg

baskaryan committed:
fmt

baskaryan committed:
fmt

baskaryan committed:
fmt

baskaryan commented:
open questions:
- should args be raw or parsed? what to do if parsing fails
- if parsed what are we doing when streaming? are we down doing partial json parsing and accumulation as part of llm stream? what if someone just wants deltas?
- does this even need to be a separate class or could we add tool_calls to AIMessage

ccurme committed:
merge master

ccurme committed:
ToolMessage --> ToolCallsMessage in openai

ccurme committed:
lint

ccurme committed:
Merge branch 'master' into bagatur/tool_calls_msg

ccurme committed:
factor out parse_partial_json and parse_tool_calls

ccurme committed:
add ToolCallsMessageChunk

ccurme committed:
add to openai

ccurme committed:
format

ccurme committed:
lint

ccurme committed:
accept None

ccurme committed:
update test_imports

ccurme committed:
ToolMessage --> ToolOutputMessage in tests

ccurme committed:
update snapshots

ccurme committed:
Merge branch 'master' into bagatur/tool_calls_msg

ccurme committed:
add ToolCall to module init

ccurme committed:
add to integration test

ccurme committed:
lint

ccurme committed:
update serializable mapping

ccurme committed:
update mistral

ccurme committed:
update test_imports

ccurme committed:
ToolOutputMessage --> ToolMessage

ccurme committed:
update snapshots

ccurme committed:
lint

ccurme committed:
clean up ToolMessage

ccurme committed:
update JsonOutputToolsParser

ccurme committed:
lint

ccurme committed:
Merge branch 'master' into bagatur/tool_calls_msg

nfcampos commented on libs/core/langchain_core/messages/tool.py:
```diff
@@ -1,26 +1,80 @@
+            ]
+
+            return self.__class__(
+                example=self.example,
```
missing `id` see other classes

baskaryan commented on libs/core/langchain_core/messages/tool.py:
```diff
@@ -1,26 +1,80 @@
+    # Ignoring mypy re-assignment here since we're overriding the value
+    # to make sure that the chunk variant can be discriminated from the
+    # non-chunk variant.
+    type: Literal["ToolCallsMessageChunk"] = "ToolCallsMessageChunk"  # type: ignore[assignment] # noqa: E501
```
does ToolCallsMessage also need it's own type to correct ser/des?

baskaryan commented on libs/core/langchain_core/messages/tool.py:
```diff
@@ -1,26 +1,80 @@
         return ["langchain", "schema", "messages"]
 
+    def __add__(self, other: Any) -> BaseMessageChunk:  # type: ignore
+        if isinstance(other, ToolCallsMessageChunk):
```
should we support merging with AIMessageChunk? and should AIMessageChunk support adding ToolCallsMesasgeChunk

baskaryan replied:
ah i guess if other is AIMessageChunk then calling super().__add__ will handle that, but think we still need to update AIMessageChunk.add to handle ToolCallsMessageChunk?

ccurme replied:
Thanks, updated this. I moved the new tool calls messages to ai.py to avoid circular imports. Now AIMessage + ToolCallsMessage = ToolCallsMessage and vice-versa.

baskaryan commented on libs/core/langchain_core/output_parsers/json.py:
```diff
@@ -44,87 +45,6 @@ def _custom_parser(multiline_string: str) -> str:
 
-# Adapted from https://github.com/KillianLucas/open-interpreter/blob/5b6080fae1f8c68938a1e4fa8667e3744084ee21/interpreter/utils/parse_partial_json.py
-# MIT License
-def parse_partial_json(s: str, *, strict: bool = False) -> Any:
```
ooc why move this? output_parsers/json still seems like a reasonable place for it, were there cyclic imports?

ccurme replied:
This was to resolve cyclic imports, but since we are no longer parsing tool calls on __add__ (since we'd need to adjust across providers) I moved this back.

baskaryan commented on libs/core/langchain_core/utils/tools.py:
```diff
@@ -0,0 +1,129 @@
+    return json.loads(s, strict=strict)
+
+
+def parse_tool_calls(
```
if this is specific to openai tools format should it still live in output_parsers/openai_tools.py? and/or have a name that makes it clear it's openai-specific

ccurme replied:
Was able to move this back to output_parsers.openai_tools since we are no longer parsing on __add__.

baskaryan commented on libs/core/langchain_core/messages/tool.py:
```diff
@@ -1,26 +1,80 @@
+    tool_calls: Optional[List[ToolCall]] = None
+
+
+class ToolCallsMessageChunk(ToolCallsMessage, BaseMessageChunk):
```
should this inherit from AIMessageChunk instead of BaseMessageChunk?

baskaryan commented on libs/core/langchain_core/messages/tool.py:
```diff
@@ -1,26 +1,80 @@
+            raw_tool_calls = additional_kwargs.get("tool_calls", [])
+            tool_calls = [
+                ToolCall(**tool_call)
+                for tool_call in parse_tool_calls(
```
hm so would this only work for openai

baskaryan replied:
i think we may want to drop this and rely on each model handling accumulating tool calls (in our format) itself, and here we just combine tool_calls lists

ccurme replied:
Updated this. For clarity, as of now when chunks are accumulated, `additional_kwargs` gets updated via `utils._merge.merge_dicts` (this may have OpenAI-specific logic in its [references to "index"](https://github.com/langchain-ai/langchain/blob/83f62fdacfa7930c975e2c4aea5733b17e8d6bde/libs/core/langchain_core/utils/_merge.py#L38), not sure if this is standard).

Tool call lists are appended, so after accumulating a few message chunks you do get things like
```
[ToolCall(name='', args={}),
 ToolCall(name='', args={}),
 ToolCall(name='', args={}),
 ToolCall(name='', args={})]
```
in `.tool_calls` in the resulting chunk (whereas the additional_kwargs looks reasonable).

During streaming, output parsers continue to parse the accumulated additional_kwargs. For `ToolCallsMessage` they can read off the parsed .tool_calls.

baskaryan reviewed (commented):
bit worried ToolCallsMessage vs ToolMessage will confuse, woudl AIMessageWithToolCalls be a silly name? AIToolCallsMessage?

ccurme committed:
Merge branch 'master' into bagatur/tool_calls_msg

ccurme committed:
cr

ccurme committed:
update ToolCallsMessageChunk.__add__

ccurme committed:
move tool calls msgs to ai

ccurme committed:
update AIMessageChunk.__add__

ccurme committed:
move parse_tool_calls

ccurme committed:
fix bug

ccurme committed:
lint

ccurme committed:
rename ToolCallsMessage

ccurme committed:
rename ToolCallsMessageChunk

ccurme committed:
lint

ccurme commented:
> bit worried ToolCallsMessage vs ToolMessage will confuse, woudl AIMessageWithToolCalls be a silly name? AIToolCallsMessage?

Updated to AIToolCallsMessage(Chunk).

ccurme committed:
merge tool calls

baskaryan committed:
fmt (#19968)

![Screenshot 2024-04-03 at 11 33 49
AM](https://github.com/langchain-ai/langchain/assets/22008038/fdf36f6a-24ec-4ab5-b008-8e2a6ffeb64a)

---------

Co-authored-by: Chester Curme <chester.curme@gmail.com>

ccurme committed:
fix bug

ccurme committed:
catch KeyError

ccurme committed:
update mistral streaming

baskaryan commented on libs/core/langchain_core/messages/ai.py:
```diff
@@ -47,6 +48,92 @@ def __add__(self, other: Any) -> BaseMessageChunk:  # type: ignore
+    type: Literal["AIToolCallsMessageChunk"] = "AIToolCallsMessageChunk"  # type: ignore[assignment] # noqa: E501
+    tool_call_chunks: Optional[List[ToolCallChunk]] = None
+
+    @classmethod
+    def get_lc_namespace(cls) -> List[str]:
+        """Get the namespace of the langchain object."""
+        return ["langchain", "schema", "messages"]
```
nit: think this is inherited

baskaryan commented on libs/partners/openai/langchain_openai/chat_models/base.py:
```diff
@@ -207,6 +238,13 @@ def _convert_delta_to_message_chunk(
         )
     elif role or default_class == ChatMessageChunk:
         return ChatMessageChunk(content=content, role=role, id=id_)
+    elif role or default_class == AIToolCallsMessageChunk:
```
think we want to check for role == "tool_calls" here

ccurme replied:
From what I've seen we enter this condition at the last chunk of a stream, where `response_metadata={'finish_reason': 'tool_calls'}`. At this point `_dict["tool_calls"]` is None, `default_class == AIToolCallsMessageChunk`, and `role` is None. This was a carry-over from when tool_calls was not optional, so I'll delete it for now. lmk if I'm misunderstanding anything.

ccurme committed:
cr

ccurme committed:
update mistral test

ccurme committed:
remove check

ccurme committed:
update docstring

ccurme committed:
add tests

ccurme commented on libs/core/langchain_core/messages/ai.py:
```diff
@@ -47,6 +48,87 @@ def __add__(self, other: Any) -> BaseMessageChunk:  # type: ignore
+    # to make sure that the chunk variant can be discriminated from the
+    # non-chunk variant.
+    type: Literal["AIToolCallsMessageChunk"] = "AIToolCallsMessageChunk"  # type: ignore[assignment] # noqa: E501
+    tool_call_chunks: Optional[List[ToolCallChunk]] = None
```
@baskaryan one issue with this: it's unclear what expected behavior for `message_chunk_to_message` should be without encoding specific parsing logic. Right now it generates an AIToolCallsMessage with tool_call_chunks. I'm not sure this is a problem though.

baskaryan replied:
could the parsing logic just be decoding json tool_call_chunks -> tool_calls?

ccurme replied:
The args in the chunks are generally substrings of json. If we wanted to do partial parsing we'd need to move the  `parse_partial_json` function.

ccurme committed:
add tests

ccurme committed:
update fireworks

ccurme committed:
update cohere and add test

ccurme committed:
add to openai tests

ccurme committed:
fix bug

ccurme committed:
Merge branch 'master' into bagatur/tool_calls_msg

ccurme committed:
update groq

ccurme committed:
update agents

baskaryan committed:
Merge branch 'master' into bagatur/tool_calls_msg

ccurme committed:
Merge branch 'bagatur/tool_calls_msg' of github.com:langchain-ai/langchain into bagatur/tool_calls_msg

ccurme committed:
Revert "update agents"

This reverts commit bc859878c11a020f6a61aee554ba66970edf8b6c.

ccurme committed:
Merge branch 'master' into bagatur/tool_calls_msg

baskaryan commented on libs/core/langchain_core/messages/ai.py:
```diff
@@ -47,6 +48,87 @@ def __add__(self, other: Any) -> BaseMessageChunk:  # type: ignore
+    # to make sure that the chunk variant can be discriminated from the
+    # non-chunk variant.
+    type: Literal["AIToolCallsMessageChunk"] = "AIToolCallsMessageChunk"  # type: ignore[assignment] # noqa: E501
+    tool_call_chunks: Optional[List[ToolCallChunk]] = None
```
i wonder if this should be Optional[Union[List[ToolCall], List[ToolCallChunk]]]

i imagine there may be some providers down the line that only stream a completed tool call (with args already as a dict), but they will yield them one at a time

ccurme replied:
`AIToolCallsMessageChunk` inherits `tool_calls`, can we put them there? That is what I did for Anthropic.

ccurme committed:
update anthropic

baskaryan commented on libs/partners/anthropic/langchain_anthropic/chat_models.py:
```diff
@@ -330,7 +333,28 @@ def _stream(
             )
-            yield cast(ChatGenerationChunk, result.generations[0])
+            message = result.generations[0].message
+            if isinstance(message, AIToolCallsMessage):
```
imo we should leave tool_call_chunks empty in this case

ccurme replied:
Having every chat model emit tool_call_chunks during streaming provides some uniformity. And this solution will work in a future where Anthropic does stream function calls. But I see how it's wasteful to `.dumps` the parsed args.

baskaryan commented on libs/partners/anthropic/langchain_anthropic/output_parsers.py:
```diff
@@ -33,11 +26,16 @@ def parse_result(self, result: List[Generation], *, partial: bool = False) -> An
+            tool_calls: List = []
+        else:
+            content: List = message.content
+            tool_calls = extract_tool_calls(content)
```
should we convert these to dicts (instead of pydantic objects). would better match our other tool parsers and pydantic objects seem overkill here. also to avoid a breaking change

ccurme replied:
fixed this, thanks

baskaryan commented on libs/partners/anthropic/langchain_anthropic/chat_models.py:
```diff
@@ -369,6 +414,12 @@ def _format_output(self, data: Any, **kwargs: Any) -> ChatResult:
         }
         if len(content) == 1 and content[0]["type"] == "text":
             msg = AIMessage(content=content[0]["text"])
+        elif any(block["type"] == "tool_use" for block in content):
```
cc @nfcampos

ccurme committed:
cr

ccurme commented on libs/core/langchain_core/messages/ai.py:
```diff
@@ -47,6 +48,87 @@ def __add__(self, other: Any) -> BaseMessageChunk:  # type: ignore
+    name: str
+    args: dict
+    id: Optional[str] = None
+    index: Optional[int] = None
```
@baskaryan how do you feel about removing `index` from ToolCall? We need it on ToolCallChunk for merging but I'm not sure it's needed on ToolCall.

baskaryan committed:
use tool call msgs (#20051)

Co-authored-by: Chester Curme <chester.curme@gmail.com>

ccurme committed:
update docstrings

ccurme committed:
spell check

ccurme commented on libs/langchain/langchain/output_parsers/json.py:
```diff
@@ -1,13 +1,7 @@
 
 __all__ = [
     "SimpleJsonOutputParser",
-    "parse_partial_json",
```
@baskaryan why not export these? isn't this breaking?

ccurme committed:
export json output parsers in langchain

ccurme committed:
undo stray import changes

ccurme committed:
tweak docstring

ccurme committed:
move tool calls to AIMessage (#20090)

ccurme committed:
merge (delete cohere)

ccurme committed:
update

ccurme committed:
merge

ccurme committed:
Merge branch 'master' into bagatur/tool_calls_msg

ccurme committed:
best effort parsing + handle parsing errors (#20111)

ccurme committed:
fix anthropic

baskaryan commented:
do we need prs in langchain-google and langchain-cohere as well?

ccurme commented:
> do we need prs in langchain-google and langchain-cohere repos as well?

yes:
- https://github.com/langchain-ai/langchain-google/pull/123
- https://github.com/langchain-ai/langchain-cohere/pull/7

ccurme committed:
cr

jacoblee93 commented on libs/core/langchain_core/messages/tool.py:
```diff
@@ -61,3 +63,105 @@ def __add__(self, other: Any) -> BaseMessageChunk:  # type: ignore
         return super().__add__(other)
+
+
+class ToolCall(Serializable):
```
How will this get serialized? `langchain.messages` or `langchain.messages.tool`?

jacoblee93 commented on libs/core/langchain_core/messages/ai.py:
```diff
@@ -16,13 +29,36 @@ class AIMessage(BaseMessage):
+            if issubclass(cls, AIMessageChunk):  # type: ignore
+                values["tool_call_chunks"] = default_tool_chunk_parser(raw_tool_calls)
+            else:
+                values["tool_calls"] = default_tool_parser(raw_tool_calls)
```
Should this check to see if `tool_calls` is passed in on `values`?

ccurme replied:
good catch!

jacoblee93 commented on libs/core/langchain_core/messages/ai.py:
```diff
@@ -35,27 +71,73 @@ class AIMessageChunk(AIMessage, BaseMessageChunk):
+                "tool_calls cannot be set on AIMessageChunk, it is derived "
+                "from tool_call_chunks."
+            )
+        if not values["tool_call_chunks"]:
```
This is a little bit confusing - why not just always overwrite to either empty array or `None`?

ccurme replied:
when summing AIMessageChunks you can get chunks with valid tool calls.

ccurme replied:
nvm got your meaning, updated this.

jacoblee93 commented on libs/core/langchain_core/messages/ai.py:
```diff
@@ -35,27 +71,73 @@ class AIMessageChunk(AIMessage, BaseMessageChunk):
+                args_ = args_ if isinstance(args_, dict) else {}
+            except (JSONDecodeError, TypeError):  # None args raise TypeError
+                args_ = {}
+            tool_calls.append(
```
Shouldn't we be combining the tool call chunks?

ccurme committed:
update tool calls to typeddict (#20208)

ccurme committed:
fix docstring

ccurme committed:
update (#20215)

ccurme committed:
fix bug

ccurme committed:
fix bug

jacoblee93 commented on libs/core/langchain_core/messages/ai.py:
```diff
@@ -16,13 +28,46 @@ class AIMessage(BaseMessage):
+    @root_validator
+    def _backwards_compat_tool_calls(cls, values: dict) -> dict:
+        raw_tool_calls = values.get("additional_kwargs", {}).get("tool_calls")
+        tool_calls = (
```
Naming here is confusing since it's shadowed below

jacoblee93 commented on libs/core/langchain_core/messages/ai.py:
```diff
@@ -16,13 +28,46 @@ class AIMessage(BaseMessage):
+                "New langchain packages are available that more efficiently handle "
+                "tool calling. Please upgrade your packages to versions that set "
+                "message tool calls. e.g., `pip install --upgrade langchain-anthropic"
+                "`, pip install--upgrade langchain-openai`, etc."
```
Missing space

jacoblee93 commented on libs/core/langchain_core/messages/ai.py:
```diff
@@ -35,27 +80,90 @@ class AIMessageChunk(AIMessage, BaseMessageChunk):
+        if not values["tool_call_chunks"]:
+            values["tool_calls"] = []
+            values["invalid_tool_calls"] = []
+            return values
```
Small nit - what if `values["tool_calls"]` is explicitly passed?

jacoblee93 commented on libs/core/langchain_core/messages/ai.py:
```diff
@@ -35,27 +80,90 @@ class AIMessageChunk(AIMessage, BaseMessageChunk):
+                args_ = parse_partial_json(chunk["args"])
+                if isinstance(args_, dict):
+                    tool_calls.append(
+                        ToolCall(
```
Don't you need `index`?

jacoblee93 commented on libs/core/langchain_core/output_parsers/openai_tools.py:
```diff
@@ -1,13 +1,89 @@
+    return parsed
+
+
+def make_invalid_tool_call(
```
I don't think this belongs in this file

jacoblee93 commented on libs/partners/groq/langchain_groq/chat_models.py:
```diff
@@ -278,9 +280,20 @@ def _stream(
+                    "id": rtc.get("id"),
+                    "index": rtc.get("index"),
+                }
+                for rtc in message.additional_kwargs["tool_calls"]
```
nit: pull from `tool_calls` on the `AIMessage`?
